src/util/rng.h

```diff
@@ -0,0 +1,64 @@
+/*********************                                                        */
+/*! \file rng.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Aina Niemetz
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief A Random Number Generator.
+ **
+ ** A random number generator, implements the xorshift* generator
+ ** (see S. Vigna, An experimental exploration of Marsaglia's xorshift
+ ** generators, scrambled. ACM Trans. Math. Softw. 42(4): 30:1-30:23, 2016).
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__RNG_H
+#define __CVC4__RNG_H
+
+namespace CVC4 {
+
+class RNG
+{
+ public:
+  RNG(RNG const&) = delete;
+  void operator=(RNG const&) = delete;
+
+  /* Get current RNG (singleton).  */
+  static RNG& getRNG()
+  {
+    static RNG s_current(0);
+    return s_current;
+  }
+
+  /* Set seed of RNG.  */
+  void setSeed(uint64_t seed)
+  {
+    d_seed = seed;
+    d_state = seed;
+  }
+
+  /* Next random uint64_t number. */
+  uint64_t rand();
+  /* Pick random uint64_t number between from and to (inclusive). */
+  uint64_t pick(uint64_t from, uint64_t to);
+  /* Pick random double number between from and to (inclusive). */
+  double pickDouble(double from, double to);
+  /* Pick with given probability (yes / no). */
+  bool pickWithProb(double probability);
+
+ private:
+  RNG(uint64_t seed) : d_seed(seed), d_state(seed) {}
```

Comment:
Just a suggestion: You might want to consider making this public. It is not necessary to enforce at the type level that this is a singleton.

Reply:
If we want to enforce that one single RNG instance is used for everything in CVC4 it is, isn't it?

Reply:
I think that enforcing having a single instance is desirable for reproducibility.

Reply:
This is just a suggestion. I have found that I personally have saved myself effort by having weaker contracts in similar situations. The payoffs were mostly when writing unit tests or encountering requirement changes.

> If we want to enforce that one single RNG instance is used for everything in CVC4 it is, isn't it?

Why do you only want to have *exactly* one? You can still have a standard shared one. I am just not sure it has to be uniquely single.

> I think that enforcing having a single instance is desirable for reproducibility.

Having multiple would not impact reproduciblity if they are all deterministic and seeded deterministically.

Reply:
I see your point. And I agree, it may be a good idea to have one shared instance but allow multiple non-shared instances. I remember tracking down performance regressions in the past where unrelated rand() calls (e.g., in unit tests) to a single RNG instance changed the sequence of random numbers encountered. However, developers must be aware that the seed set via options only influences the shared RNG instance.